### PR TITLE
Merkle fixes

### PR DESCRIPTION
**Pull request by bobbinth**
This PR contains a few updates to the Merkle struct. Specifically:

- We now require that `NodeIndex` always exists in a valid state. The `new()` constructor now returns an error and there is a special constructor for tests (called `make()`) which encapsulates error unwrapping.
- `BItIterator` struct was removed. The only place where it was being used was 2 functions in the `MerkleStore` and I don't think the logic there became any more complicated.
- A few bugs were fixed in benchmarks and `MerkleStore` code.

This PR will supersede #92 (I copied tests and some comments from there).

### CODE REVIEW

**Review by vlopes11 — APPROVED** [review]
Looks good! I adapted the test case that originated this issue to a correct version that could be optionally added to further improve the robustness of the implementation

```rust
#[test]
fn node_path_should_be_truncated() {
    let key = 0b11010010_11001100_11001100_11001100_11001100_11001100_11001100_11001100;

    let mut store = MerkleStore::new();
    let root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();

    // insert first node - works as expected
    let depth = 64;
    let node = [Felt::new(key); WORD_SIZE];
    let index = NodeIndex::new(depth, key).unwrap();
    let root = store.set_node(root, index, node).unwrap().root;
    let result = store.get_node(root, index).unwrap();
    let path = store.get_path(root, index).unwrap().path;
    assert_eq!(node, result);
    assert_eq!(path.depth(), depth);
    assert!(path.verify(index.value(), result, &root));

    // flip the first bit of the key and insert the second node on a different depth
    let key = key ^ (1 << 63);
    let key = key >> 8;
    let depth = 56;
    let node = [Felt::new(key); WORD_SIZE];
    let index = NodeIndex::new(depth, key).unwrap();
    let root = store.set_node(root, index, node).unwrap().root;
    let result = store.get_node(root, index).unwrap();
    let path = store.get_path(root, index).unwrap().path;
    assert_eq!(node, result);
    assert_eq!(path.depth(), depth);
    assert!(path.verify(index.value(), result, &root));

    // attempt to fetch a path of the second node to depth 64
    let key = key << 8;
    let index = NodeIndex::new(64, key).unwrap();
    assert!(store.get_node(root, index).is_err());
}
```

File: src/merkle/simple_smt/mod.rs [hunk 1]
```diff
@@ -119,12 +119,12 @@ impl SimpleSmt {
...
     }
 
-    /// Returns a node at the specified key
+    /// Returns a node at the specified index.
     ///
     /// # Errors
     /// Returns an error if:
     /// * The specified depth is greater than the depth of the tree.
-    pub fn get_node(&self, index: &NodeIndex) -> Result<Word, MerkleError> {
+    pub fn get_node(&self, index: NodeIndex) -> Result<Word, MerkleError> {
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply src/merkle/simple_smt/mod.rs:127]: I believe this was changed due to this conversation: https://github.com/0xPolygonMiden/crypto/pull/121#discussion_r1156238948

**Comment by hackaugusto** [comment]
Just to document my two cents on the PR, IMO the API user shouldn't have to do the bitshifts, using @vlopes11 example above:

```rust
    let key = key ^ (1 << 63);
    let key = key >> 8;
    let depth = 56;
```

The solution is to define the stable iteration for the root-to-leaf case, instead of the current version without the bitshift that is from the leaf-to-root.

File: src/bit.rs [hunk 2]
```diff
@@ -1,169 +0,0 @@
-/// Yields the bits of a `u64`.
-pub struct BitIterator {
-    /// The value that is being iterated bit-wise
-    value: u64,
-    /// True bits in the `mask` are the bits that have been visited.
-    mask: u64,
-}
-
-impl BitIterator {
```
Comment (hackaugusto) [inline src/bit.rs:9]: question: this is basically an iterator API for the `NodeIndex`, so what I got from this is that we don't want to have an iterator for it?

I think that having an iterator is easier to understand that have state mutation on the loop body.

File: src/merkle/index.rs [hunk 3]
```diff
@@ -19,33 +29,44 @@ impl NodeIndex {
...
+        }
+    }
+
+    /// Creates a new node index for testing purposes.
+    ///
+    /// # Panics
+    /// Panics if the `value` is greater than or equal to 2^{depth}.
+    #[cfg(test)]
+    pub fn make(depth: u8, value: u64) -> Self {
+        Self::new(depth, value).unwrap()
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply src/merkle/index.rs:49]: Yep - this makes sense. I guess in this case the question is whether we need `make()` constructor. I kept it mainly for brevity - but semantically `new_unchecked()` and `make()` would be the same.

File: src/merkle/store/mod.rs [hunk 4]
```diff
@@ -353,24 +360,19 @@ impl MerkleStore {
...
-                roots.iter().map(|v| Word::from(*v)).collect(),
-            ));
-        }
-
-        for (index_value, node, path) in paths {
-            self.add_merkle_path(index_value, node, path)?;
+        let mut root = Word::default();
+        for (index_value, node, path) in paths.into_iter() {
+            let new_root = self.add_merkle_path(index_value, node, path)?;
+            if new_root != root {
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply src/merkle/store/mod.rs:366]: Good point! Now that I think about it more I wonder if we need this restriction at all. I'm not sure there is a good reason to enforce that all paths must resolve to the same root.

[the remark this answers is not in the document]
Comment (bobbinth) [reply src/merkle/store/mod.rs]: I've removed the requirement that all paths must resolve to the same root from here.

File: src/bit.rs (hunk 2 above)
Comment (bobbinth) [reply src/bit.rs:9]: I think for now it is simpler without it. The iterator was like 100+ lines of code which, in my opinion, made what's going on non-obvious. I replaced it with something like 4 lines of code which I think are not too difficult to interpret.

We can of course come back to it later on if we can use it more widely.

**Pull request merged**